api/comms/rpc_processor.go

```diff
@@ -0,0 +1,352 @@
+			}
+		case RPCMethodChatMessage:
+			var params ChatMessageRPCParams
+			err = json.Unmarshal(rawRpc.Params, &params)
+			if err != nil {
+				return err
+			}
+			err = chatSendMessage(tx, ctx, userId, params.ChatID, params.MessageID, messageTs, params.Message)
+			if err != nil {
+				return err
+			}
+		case RPCMethodChatReact:
+			var params ChatReactRPCParams
+			err = json.Unmarshal(rawRpc.Params, &params)
+			if err != nil {
+				return err
+			}
+			err = chatReactMessage(tx, ctx, userId, params.ChatID, params.MessageID, params.Reaction, messageTs)
+			if err != nil {
+				return err
+			}
+
+		case RPCMethodChatRead:
+			var params ChatReadRPCParams
+			err = json.Unmarshal(rawRpc.Params, &params)
+			if err != nil {
+				return err
+			}
+
+			// do nothing if last active at >= message timestamp
+			var lastActive pgtype.Timestamp
+			const lastActiveAtQuery = `
+select last_active_at from chat_member where chat_id = $1 and user_id = $2`
+
+			err = tx.QueryRow(ctx, lastActiveAtQuery, params.ChatID, userId).Scan(&lastActive)
+			if err != nil {
+				return err
+			}
+
+			if !lastActive.Valid || messageTs.After(lastActive.Time) {
+				err = chatReadMessages(tx, ctx, userId, params.ChatID, messageTs)
+				if err != nil {
+					return err
+				}
+			}
+		case RPCMethodChatPermit:
+			var params ChatPermitRPCParams
+			err = json.Unmarshal(rawRpc.Params, &params)
+			if err != nil {
+				return err
+			}
+			err = chatSetPermissions(tx, ctx, userId, params.Permit, params.PermitList, params.Allow, messageTs)
+			if err != nil {
+				return err
+			}
+		case RPCMethodChatBlock:
+			var params ChatBlockRPCParams
+			err = json.Unmarshal(rawRpc.Params, &params)
+			if err != nil {
+				return err
+			}
+			blockeeUserId, err := trashid.DecodeHashId(params.UserID)
+			if err != nil {
+				return err
+			}
+			err = chatBlock(tx, ctx, userId, int32(blockeeUserId), messageTs)
+			if err != nil {
+				return err
+			}
+		case RPCMethodChatUnblock:
+			var params ChatUnblockRPCParams
+			err = json.Unmarshal(rawRpc.Params, &params)
+			if err != nil {
+				return err
+			}
+			unblockedUserId, err := trashid.DecodeHashId(params.UserID)
+			if err != nil {
+				return err
+			}
+			err = chatUnblock(tx, ctx, userId, int32(unblockedUserId), messageTs)
+			if err != nil {
+				return err
+			}
+
+		case RPCMethodChatBlast:
+			var params ChatBlastRPCParams
+			err = json.Unmarshal(rawRpc.Params, &params)
+			if err != nil {
+				return err
+			}
+
+			outgoingMessages, err := chatBlast(tx, ctx, userId, messageTs, params)
+			if err != nil {
+				return err
+			}
+			// Send chat message websocket event to all recipients who have existing chats
+			for _, outgoingMessage := range outgoingMessages {
+				_, err := json.Marshal(outgoingMessage.ChatMessageRPC)
+				if err != nil {
+					logger.Error("err: invalid json", zap.Error(err))
+				} else {
+					// TODO
+					// websocketNotify(json.RawMessage(j), userId, messageTs.Round(time.Microsecond))
```

Comment:
plz dont flip prod over to this until we have this sorted! will break chat blasts

Reply:
Oh yeah don't sweat, there are at least two additional PRs to finish migrating functionality before I'm even going to turn this on in staging 😅